### ISSUE TITLE
Configure cluster network cidr blocks

### ISSUE THREAD

**Issue by jschoone** [SovereignCloudStack/k8s-cluster-api-provider#419]
As an SCS user, I want to be able to configure the network CIDR blocks for my Kubernetes clusters created by ClusterAPI, so that I can define the IP address range for the pods and services in my cluster.

**Comment by jschoone** [comment]
> > What is needed beyond node_cidr, NODE_CIDR documented in https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/doc/configuration.md ?
> 
> I think it is all about customizable clusterNetwork.pods.[cidrBlocks](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/terraform/files/template/cluster-template.yaml#L10) - nodeCidr is used [here](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/terraform/files/template/cluster-template.yaml#L36)

Correct. Besides the Pod network also [clusterNetwork.services](https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/Cluster/v1beta1@v1.4.2#spec-clusterNetwork-services) can be configured.

**Comment by fdobrovolny** [comment]
# Pod CIDR Issue

Based on the #454 PR review, it was discovered that Cilium, by default, does not respect the setting from CAPI, and Calico may respect it only partially. I want to summarise the findings and discuss my proposed solution here.

## Cilium
Cilium uses two different IPAM modes: [Kubernetes Host Scope](https://docs.cilium.io/en/stable/network/concepts/ipam/kubernetes/) and [Cluster Pool](https://docs.cilium.io/en/stable/network/concepts/ipam/cluster-pool/) (default). In the Kubernetes Host Scope mode, Cilium delegates the address allocation to each individual node in the cluster, with IPs allocated out of the PodCIDR range associated with each node by Kubernetes. In the Cluster Pool mode, Cilium assigns per-node PodCIDRs to each node and allocates IPs using a host-scope allocator on each node. The pool from which the IP's are allocated is specified in helm during configuration and is not derived from the value provided by CAPI. The difference is that instead of Kubernetes assigning the per-node PodCIDRs via the Kubernetes `v1.Node` resource, the Cilium operator manages the per-node PodCIDRs via the `v2.CiliumNode` resource. This mode does not depend on Kubernetes being configured to hand out per-node PodCIDRs. However, in our current setup, the cluster is configured to do so by specifying the pod CIDR.

## Calico

https://docs.tigera.io/calico/latest/networking/ipam/get-started-ip-addresses

Calico provides its own IPAM plugin called calico-ipam, which uses Calico’s IP pool resource to control how IP addresses are allocated to pods within the cluster. By default, Calico uses a single IP pool for the entire Kubernetes pod CIDR, but you can divide the pod CIDR into several pools. The IP pool used by default is the one specified in CAPI. Calico also supports the host-local IPAM plugin, which uses predetermined CIDRs statically allocated to each node by Kubernetes to choose addresses for pods. However, some Calico features are unavailable when using the host-local IPAM plugin, such as per namespace IPPool.

## Summary

The host-local IPAM plugin and the Kubernetes Host Scope work on the same principle.

In the current setup, Cilium is ignoring the pod CIDR range set up in Kubernetes via CAPI, and Calico is using it as a whole and not adhering to the per-node CIDR ranges handed out by Kubernetes.

Given this situation, I propose using Kubernetes-issued per-node CIDRs. This would involve configuring Cilium to use the Kubernetes Host Scope mode, which delegates the address allocation to each individual node in the cluster, with IPs allocated out of the PodCIDR range associated with each node by Kubernetes. I recommend using the host-local IPAM plugin for Calico to have consistent behaviour across Cilium and Calico.



# Backwards compatibility

As PR #454 affects both `clusterctl.yaml` and `cluster-template.yml`, a way to update existing management nodes has to be created. In the past, I found this has been done in two ways.

1. Updating Upgrade-Guide and letting the user fix the files by themselves. - 3abfa31
2. Creating a script which would patch the existing template. - e4ae4fbc86f93b06744fa7acf747d32805bebc41

Here I want to propose a more systematic change which could prevent this issue from happening in the future, and I will write this in a separate issue. However, what do you guys recommend as the approach I should take for now?

**Comment by fdobrovolny** [comment]
After discussing with @matofeder, we decided on his proposed solution, which I have successfully implemented it in the PR #454 .
